Provide more information on args in traces

kdambekalns opened the pull request:
This provides a better representation of function arguments in traces that are objects. So far only the type of the object was visible, now a string or JSON representation is shown, if possible.

kdambekalns committed:
Provide more information on args in traces

This provides a better representation of function arguments in traces
that are objects. So far only the type of the object was visible, now
a string or JSON representation is shown, if possible.

kdambekalns commented:
Before:

<img width="918" alt="Screenshot 2022-09-09 at 10 44 43" src="https://user-images.githubusercontent.com/95873/189322171-c53e8697-0251-4bf1-9cbf-3e024185e7f6.png">

After:

<img width="903" alt="Screenshot 2022-09-09 at 11 35 03" src="https://user-images.githubusercontent.com/95873/189322195-b7f4ff0e-88a4-44a0-8623-85a3b4111a0d.png">

kdambekalns commented:
Hm, since `type` is always a class name, it could be `class` instead…

robertlemke commented on Classes/RepresentationSerializer.php:
```diff
@@ -0,0 +1,52 @@
+/*
+ * This file is part of the Flownative.Sentry package.
+ *
+ * (c) Karsten Dambekalns, Flownative GmbH - www.flownative.com
```
I think we should find some consistent copyright notice, as mentioning the author per file probably doesn't even have the desired legal effect.

I suggest that we either keep on using `(c) Robert Lemke, Flownative GmbH´ to give attribution to me being the original author or we use something more general.

Certainly not comparable, but just for inspiration, I looked at Phpunit, for example, which only mentions Sebastian as the original author despite the contributions by > 400 people.

I don't mean to diminish your contribution though, so if it's important to you, let's add more names.

kdambekalns replied:
Whatever… I think the names are more or less useless TBH. But since you didn't create that file, it felt weird to put your name there. 🤷‍♂️

robertlemke commented on Classes/Command/SentryCommandController.php:
```diff
@@ -66,16 +71,6 @@ public function testCommand(): void
-    {
-        $previousException = new \RuntimeException('Test "previous" exception thrown by the SentryCommandController', 1614759554);
-        throw new SentryClientTestException('Test exception in SentryCommandController', 1614759519, $previousException);
+        (new ThrowingClass())->throwException(new StringableTestArgument((string)M_PI));
```
Why does that need to be in braces? Did you mean this:

```php
new ThrowingClass())->throwException(new StringableTestArgument((string)M_PI);
```

kdambekalns replied:
It's used instead of:

```php
$c = new ThrowingClass();
$c->throwException(new StringableTestArgument((string)M_PI));
```

`(new ThrowingClass())` is needed to avoid `Parse error: syntax error, unexpected token "->" in`, you cannot do

```php
new ThrowingClass()->throwException(new StringableTestArgument((string)M_PI);
```

robertlemke replied:
I see – never stumbled over that, probably because you usually assign the new object to some variable.

kdambekalns committed:
Unify class docblocks

kdambekalns committed:
Adjust index keys in serialized values

An unchanged object reads "Object \Some\Class\Name", so this changes
the serialized representation to use "Object" instead of "type".

kdambekalns commented:
> Hm, since `type` is always a class name, it could be `class` instead…

Changed it to `Object`…

robertlemke commented:
I tested this locally. It worked fine for the `sentry:test` command, but so far it did not work when I caused real errors in the application.

<img width="1405" alt="image" src="https://user-images.githubusercontent.com/95582/189361266-0ae604be-9167-46f9-9618-9759627cb4c8.png">

The value object in question does provide a `jsonSerialize()` and `__toString()` method.

kdambekalns commented:
Ok, I'll have a closer look. 🔎

kdambekalns commented:
That `{serialization error}` actually is something like this: `Cannot acquire reference to readonly property Flownative\Beach\Domain\ValueObject\AbstractIdentifier::$uuid`

kdambekalns committed:
Fix "serialization error" caused by reference usage

The parent `serializeObject()` method used assign-by-ref when iterating
over the `$object`. With a readonly property that fails…

kdambekalns committed:
Add more specialized class serializers

-

kdambekalns commented:
Ping…

robertlemke committed:
Merge branch 'main' into feature/readable-backtrace-vars

robertlemke committed:
Tweak representation of serialized arguments

robertlemke commented:
Thanks! 

Are you fine with this little modification, which moves the "data" one level up as first-level properties of the JSON object? The actual properties can then be distinguished from our "class" property by prefixing it as "__class". Makes it a bit easier to read for my taste, because you don't have to expand nested objects in Sentry in order to see what's inside.

<img width="1088" alt="image" src="https://user-images.githubusercontent.com/95582/220321850-21e24410-e77e-4778-87d2-76ea188cd4f4.png">
